File: dental/dental/G00/Model/LoginBean.swift
//
//  LoginBean.swift
//  dental
//
//  Created by SPJ on 1/9/18.
//  Copyright © 2018 SPJ. All rights reserved.
//

import UIKit
import harpyframework

class LoginBean: ConfigBean {
    /**
     * Object instance
     */
    public static let shared: LoginBean = {
        let instance = LoginBean()
        return instance
    }()
    // MARK: Properties
    /** User token */
    var token:              String          = DomainConst.BLANK
    /** Id of role */
    var role_id:            String          = DomainConst.BLANK
    /** Flag need change pass */
    var need_change_pass:   String          = DomainConst.BLANK
    /** List menu */
    var menu:               [ConfigBean]    = [ConfigBean]()
    /** Pathological config */
    var pathological:       [ConfigBean]    = [ConfigBean]()
    /** Address config */
    var address_config:     [ConfigBean]    = [ConfigBean]()
    /** Diagnosis config */
    var diagnosis:          [ConfigBean]    = [ConfigBean]()
    /** Treatment type config */
    var treatment:          ListConfigBean  = ListConfigBean()
    /** Teeth config */
    var teeth:              [ConfigBean]    = [ConfigBean]()
    
    override public init() {
        super.init()
    }
    
    /**
     * Initializer
     * - parameter jsonData: List of data
     */
    public override init(jsonData: [String: AnyObject]) {
        super.init()
        // Token
        self.token = getString(json: jsonData, key: DomainConst.KEY_TOKEN)
        // Role id
        self.role_id = getString(json: jsonData, key: DomainConst.KEY_ROLE_ID)
        // Flag need change password
        self.need_change_pass = getString(json: jsonData, key: DomainConst.KEY_NEED_CHANGE_PASS)
        // Menu
        self.menu.append(contentsOf: getListConfig(json: jsonData, key: DomainConst.KEY_MENU))
        // Pathological
        self.pathological.append(contentsOf: getListConfig(json: jsonData, key: DomainConst.KEY_PATHOLOGICAL))
        self.address_config.append(contentsOf: getListConfig(json: jsonData, key: DomainConst.KEY_ADDRESS_CONFIG))
        // Diagnosis
        self.diagnosis.append(contentsOf: getListConfig(json: jsonData, key: DomainConst.KEY_DIAGNOSIS))
        // Treatment
        if let dataArr = jsonData[DomainConst.KEY_TREATMENT] as? [[String: AnyObject]] {
            self.treatment = ListConfigBean.init(jsonData: dataArr)
        }
        // Teeth
        self.teeth.append(contentsOf: getListConfig(json: jsonData, key: DomainConst.KEY_TEETH))
    }
    
    /**
     * Get city by id
     * - parameter id:      Id of city
     * - returns: City model
     */
    public func getCityById(id: String) -> ConfigBean {
        for item in self.address_config {
            if item.id == id {
                return item
            }
        }
        return ConfigBean()
    }
    
    /**
     * Get district by id
     * - parameter id:          Id of district
     * - parameter cityId:      Id of city
     * - returns: District model
     */
    public func getDistrictById(id: String, cityId: String) -> ConfigBean {
        for item in self.getCityById(id: cityId).data {
            if item.id == id {
                return item
            }
        }
        return ConfigBean()
    }
    
    /**
     * Get ward by id
     * - parameter id:          Id of ward
     * - parameter cityId:      Id of city
     * - parameter districtId:  Id of district
     * - returns: Ward model
     */
    public func getWardById(id: String, cityId: String, districtId: String)
        -> ConfigBean {
            for item in self.getDistrictById(id: districtId, cityId: cityId).data {
                if item.id == id {
                    return item
                }
            }
            return ConfigBean()
    }
    
    /**
     * Get diagnosis config value
     * - parameter id:          Id of config
     */
    public func getDiagnosisConfig(id: String) -> String {
        for item in self.diagnosis {
            if item.id == id {
                return item.name
            }
        }
        return DomainConst.BLANK
    }
        
    /**
     * Get treatment config value
     * - parameter id: Id of config
     */
    public func getTreatmentConfig(id: String) -> String {
        for item in self.treatment.getData() {
            for child in item._dataExt {
                if child.id == id {
                    return child.name
                }
            }
        }
        return DomainConst.BLANK
    }
    
    /**
     * Get teeth config value
     * - parameter id:          Id of config
     */
    public func getTeethConfig(id: String) -> String {
        for item in self.teeth {
            if item.id == id {
                return item.name
            }
        }
        return DomainConst.BLANK
    }
    
    /**
     * Get update text string
     * - returns: String
     */
    public func getUpdateText() -> String {
        return DomainConst.CONTENT00555
    }
}

File: dental/app/src/main/java/vietmydental/immortal/com/gate/g00/model/LoginBean.java
package vietmydental.immortal.com.gate.g00.model;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import org.json.JSONObject;

import java.util.ArrayList;

import vietmydental.immortal.com.gate.model.ConfigBean;
import vietmydental.immortal.com.gate.model.ConfigExtBean;
import vietmydental.immortal.com.gate.utils.CommonProcess;
import vietmydental.immortal.com.gate.utils.DomainConst;

public class LoginBean extends ConfigBean {
    /** Singleton instance */
    private static LoginBean instance;
    // MARK: Properties
    /** User token */
    private String token                            = DomainConst.BLANK;
    /** Id of role */
    private String role_id                          = DomainConst.BLANK;
    /** Flag need change pass */
    private String need_change_pass                 = DomainConst.BLANK;
    /** List menu */
    private ArrayList<ConfigBean> menu              = new ArrayList<>();
    /** Pathological config */
    private ArrayList<ConfigBean> pathological      = new ArrayList<>();
    /** Address config */
    private ArrayList<ConfigBean> address_config    = new ArrayList<>();
    /** Diagnosis config */
    private ArrayList<ConfigExtBean> diagnosis      = new ArrayList<>();
    /** Treatment type config */
    private ArrayList<ConfigExtBean> treatment      = new ArrayList<>();
    /** Teeth config */
    private ArrayList<ConfigExtBean> teeth          = new ArrayList<>();
    /** Timer config */
    private ArrayList<ConfigBean> timer             = new ArrayList<>();

    /**
     * Constructor.
     */
    public LoginBean() {
        super();
    }

    /**
     * Get singleton instance.
     * @return Singleton instance
     */
    public static LoginBean getInstance() {
        if (instance == null) {
            instance = new LoginBean();
        }
        return instance;
    }

    /**
     * Constructor
     * @param data JSONObject data
     */
    public void updateData(JSONObject data) {
        JsonParser jsonParser = new JsonParser();
        JsonObject gsonObject = (JsonObject) jsonParser.parse(data.toString());
        this.id         = CommonProcess.getJsonString(gsonObject, DomainConst.KEY_ID);
        this.name       = CommonProcess.getJsonString(gsonObject, DomainConst.KEY_NAME);
        this.token              = CommonProcess.getJsonString(gsonObject, DomainConst.KEY_TOKEN);
        this.role_id            = CommonProcess.getJsonString(gsonObject, DomainConst.KEY_ROLE_ID);
        this.need_change_pass   = CommonProcess.getJsonString(gsonObject, DomainConst.KEY_NEED_CHANGE_PASS);
        this.menu = CommonProcess.getListConfig(gsonObject, DomainConst.KEY_MENU);
        this.menu.add(new ConfigBean(DomainConst.MENU_ID_LIST.CONFIGURATION, "Cài đặt"));
        this.menu.add(new ConfigBean(DomainConst.MENU_ID_LIST.LOGOUT, "Thoát"));
        this.pathological       = CommonProcess.getListConfig(gsonObject, DomainConst.KEY_PATHOLOGICAL);
        this.address_config     = CommonProcess.getListConfig(gsonObject, DomainConst.KEY_ADDRESS_CONFIG);
        this.diagnosis          = CommonProcess.getListConfigExt(gsonObject, DomainConst.KEY_DIAGNOSIS);
        this.treatment          = CommonProcess.getListConfigExt(gsonObject, DomainConst.KEY_TREATMENT);
        this.teeth              = CommonProcess.getListConfigExt(gsonObject, DomainConst.KEY_TEETH);
        this.timer              = CommonProcess.getListConfig(gsonObject, DomainConst.KEY_TIMER);
    }

    /**
     * Clear data when logout.
     */
    public void clearData() {
        this.token = DomainConst.BLANK;
        this.role_id = DomainConst.BLANK;
        this.menu.clear();
        this.menu.add(new ConfigBean(DomainConst.MENU_ID_LIST.LOGIN, "Đăng nhập"));
        this.menu.add(new ConfigBean(DomainConst.MENU_ID_LIST.CONFIGURATION, "Cài đặt"));
        this.getPathological().clear();
        this.address_config.clear();
        this.getDiagnosis().clear();
        this.treatment.clear();
        this.teeth.clear();
        this.timer.clear();
    }

    /**
     * Get token value
     * @return Token value
     */
    public String getToken() {
        return this.token;
    }

    /**
     * Get menu value
     * @return Menu value
     */
    public ArrayList<ConfigBean> getMenu() {
        return menu;
    }

    /**
     * Get pathological value
     * @return Pathological value
     */
    public ArrayList<ConfigBean> getPathological() {
        return pathological;
    }
    public ArrayList<ConfigBean> getTimer() {
        return timer;
    }

    public ArrayList<ConfigExtBean> getDiagnosis() {
        return diagnosis;
    }

    /**
     * Get list full of diagnosis, only children items in 1 array
     * @return Children list items
     */
    public ArrayList<ConfigExtBean> getDiagnosisFull() {
        ArrayList<ConfigExtBean> retVal = new ArrayList<>();
        for (ConfigExtBean item :
                diagnosis) {
            for (ConfigExtBean child :
                    item.getDataExt()) {
                retVal.add(child);
            }
        }
        return retVal;
    }

    public ArrayList<ConfigExtBean> getTooth() {
        return teeth;
    }

    public ArrayList<ConfigExtBean> getTreatment() {
        return treatment;
    }

    /**
     * Get diagnosis by id
     * @param id Id of item
     * @return Object
     */
    public ConfigExtBean getDiagnosis(String id) {
        for (ConfigExtBean item :
                diagnosis) {
            if (id.equals(item.getId())) {
                return item;
            } else {
                if (!item.getDataExt().isEmpty()) {
                    for (ConfigExtBean child :
                            item.getDataExt()) {
                        if (id.equals(child.getId())) {
                            return child;
                        }
                    }
                }
            }
        }
        return new ConfigExtBean();
    }

    /**
     * Get treatment by id
     * @param id Id value
     * @return Object
     */
    public ConfigExtBean getTreatment(String id) {
        for (ConfigExtBean item :
                treatment) {
            for (ConfigExtBean child :
                    item.getDataExt()) {
                if (child.getId().equals(id)) {
                    return child;
                }
            }
        }
        return null;
    }
}
